### PR TITLE
Ensure regen utility class gets marked as done when concretised

### PR DESCRIPTION
**Pull request by maybe-sybr**
## Description

Ensure `regen()` instances get marked as done when concretised via the
`UserList.data` property implementation. We also improve laziness of
`regen.__repr__()` so it reflects the fact that it's lazy.

Fixes: #6786

### CODE REVIEW

File: t/unit/utils/test_functional.py [hunk 1]
```diff
@@ -150,6 +152,56 @@ def build_generator():
...
+    def test_repr(self):
+        def die():
+            raise AssertionError("Generator died")
+            yield None
+
+        # Confirm that `regen()` instances are not concretised when represented
+        g = regen(die())
+        assert "..." in repr(g)
+
+    @pytest.mark.xfail(raises=StopIteration, reason="XXX NEEDS AN ISSUE")
```
Comment (maybe-sybr) [inline t/unit/utils/test_functional.py:173]: This test is for an edge case where a repeatable iterator could raise an exception at some point before completion which would result in `regen()` stashing the elements it saw but never marking itself as done. Then the next time we yield elements from an iterator derived from the `regen()` instance, we'd yield the stashed elements, then repeat them from the underlying iterator if it's capable of giving them to us again.

IDK if this is something we can even bother handling. If the underlying iterator explodes, we could choose to mark the regen instance as `__done` and just give up on attempting to get more elements? If we want to go down that route, I can roll a change into this PR since it'd be pretty simple and still thematically relevant. If we think we need to do something more involved or optimistic, I'll make a follow up issue for discussion.

File: celery/utils/functional.py [hunk 2]
```diff
@@ -242,11 +242,20 @@ def __bool__(self):
     @property
     def data(self):
         try:
-            self.__consumed.extend(list(self.__it))
+            if not self.__done:
+                self.__consumed.extend(list(self.__it))
+            self.__done = True
```
Comment (Suor) [inline celery/utils/functional.py:247]: Might be movedunder `if`
```suggestion
                self.__done = True
```

**Comment by maybe-sybr** [comment]
> Why do we have `except StopIteration` here at all? `list()` should handle that.

Good question. That was added in 99827730 and looks like it could have been deliberate, but I have no idea why that might be the case. AFAICT you'd have to be raising `StopIteration` from inside a generator (which is wrong) to need something like that :/ A quick sanity check of `list(iter(range(5)))` works without raising and, as expected, spits out `StopIteration` when I `next()` the iterator manually. Not sure what the point is there.

Removing the `try: except StopIteration:` doesn't seem to break any unit or integration tests (against redis), so I'll amend the diff and re-push shortly.

File: celery/utils/functional.py (hunk 2 above)
Comment (maybe-sybr) [reply celery/utils/functional.py:246]: The newly pushed diff makes this change due to the dedent of the rest of the try body

File: t/unit/utils/test_functional.py [hunk 3]
```diff
@@ -150,6 +152,56 @@ def build_generator():
...
+                self._errored = False
+
+            def __iter__(self):
+                yield from self.iter_
+                if not self._errored:
+                    try:
+                        # This should stop the regen instance from marking
+                        # itself as being done
+                        #raise AssertionError("Iterator errored")
+                        pass
```
Comment (maybe-sybr) [inline t/unit/utils/test_functional.py:187]: whoops, this needs to be uncommented

Comment (maybe-sybr) [reply t/unit/utils/test_functional.py]: done

File: t/unit/utils/test_functional.py (hunk 1 above)
Comment (maybe-sybr) [reply t/unit/utils/test_functional.py]: I've made #6794 to track the discussion on this point. Could just be a thing we need to document, happy to have this PR merged without addressing that issue yet.

File: celery/utils/functional.py [hunk 4]
```diff
@@ -241,12 +241,18 @@ def __bool__(self):
 
     @property
     def data(self):
-        try:
+        if not self.__done:
             self.__consumed.extend(list(self.__it))
```
Comment (Suor) [inline celery/utils/functional.py:245]: An intermediate `list()` should not be needed. I.e. `self.__consumed.extend(self.__it)` should do.

**Comment by Suor** [comment]
> AFAICT you'd have to be raising StopIteration from inside a generator (which is wrong) to need something like that

Since Python 3.7 raising `StopIteration` inside generator results in a `RuntimeError`.

**Comment by thedrow** [comment]
> 
> 
> > AFAICT you'd have to be raising StopIteration from inside a generator (which is wrong) to need something like that
> 
> Since Python 3.7 raising `StopIteration` inside generator results in a `RuntimeError`.

This is probably an oversight we made during the Python 3 only migration in 5.0.
Thanks for noticing 😄

File: celery/utils/functional.py (hunk 4 above)
Comment (maybe-sybr) [reply celery/utils/functional.py]: changed

**Comment by maybe-sybr** [comment]
I just pushed a change which removed what I thought was a pointless `iter(self)` but I now realise that it was to ensure that the `__done` attribute gets set when the underlying iterator is completed without having to do hard stuff in the `__getitem__()` method. I've stripped that commit out, but I still think it's wasteful to spend time iterating on the (potentially many) initial elements of ourself just to get some subsequent element from the underlying iterator.

**Comment by Suor** [comment]
> I just pushed a change which removed what I thought was a pointless iter(self) but I now realise that it was to ensure that the __done attribute gets set when the underlying iterator is completed without having to do hard stuff in the __getitem__() method. I've stripped that commit out, but I still think it's wasteful to spend time iterating on the (potentially many) initial elements of ourself just to get some subsequent element from the underlying iterator.

I wonder how much use that `.__getitem__()` call gets. At least for `.map()` and friends we simply serialize, i.e. convert to list the thing eventually, we only need `regen()` to not consume immediately, since a signature might never get sent.

**Comment by maybe-sybr** [comment]
> I wonder how much use that `.__getitem__()` call gets. At least for `.map()` and friends we simply serialize, i.e. convert to list the thing eventually, we only need `regen()` to not consume immediately, since a signature might never get sent.

I'd guess the only time it really gets called is to access the 0th or `-1`th item which are used for various things in the canvas code. Otherwise I'd expect it to be mostly simple `for x in some_regen_thing:` usage.

File: t/unit/utils/test_functional.py [hunk 5]
```diff
@@ -150,6 +152,60 @@ def build_generator():
...
+        # If the `regen()` instance doesn't think it's done then it'll dupe the
+        # elements from the underlying iterator if it can be re-used
+        iter_g = iter(g)
+        for e in original_list * 2:
+            assert next(iter_g) == e
+        with pytest.raises(StopIteration):
+            next(iter_g)
+        assert getattr(g, "_regen__done") is True
+        # Finally we xfail this test to keep track of it
+        raise pytest.xfail(reason="#6794")
```
Comment (thedrow) [inline t/unit/utils/test_functional.py:207]: Why do we raise this instead of marking it as xfail?
xfailed tests still run.

Comment (maybe-sybr) [reply t/unit/utils/test_functional.py:207]: The test is verifying the incorrect behaviour in full before xfailing itself, rather than relying on some assertion failing partway through and marking it as an xfailure. I chose to do it that way because it's not clear to me how we should be handling situations like this so I can't write the "correct" behaviour at the moment. I'm sort of expecting the test to start hard failing if/when we work out how to treat something like this, or alternatively we simply leave it as an expected edge case and remove the `raise xfail()` on this line.

**Review by auvipy — APPROVED** [review]
let's merge it after some 5.1.x point releases

**Comment by thedrow** [comment]
> 
> 
> let's merge it after some 5.1.x point releases

I think we can include it in 5.1.1 unless there's a special reason not to do so that I missed.

**Comment by auvipy** [comment]
OK. merged as bugfix
